skbio/diversity/beta/_unifrac.py

```diff
@@ -189,44 +176,325 @@ def weighted_unifrac(u_counts, v_counts, otu_ids, tree, normalized=False,
+                      v_total_count, branch_lengths):
+    """
+    Parameters
+    ----------
+    u_node_counts, v_node_counts : np.array
+        Vectors indicating presense (value greater than zero) and absense
+        (value equal to zero) of nodes in two samples, `u` and `v`. Order is
+        assumed to be the same as in `branch_lengths`.
+    u_total_count, v_total_counts : int
+        The sum of ``u_node_counts`` and ``v_node_counts`` vectors,
+        respectively. This could be computed internally, but since this is a
+        private method and the calling function has already generated these
+        values, this saves an iteration over each of these vectors.
+    branch_lengths : np.array
+        Vector of branch lengths of all nodes (tips and internal nodes) in
+        postorder representation of their tree.
+
+    Returns
+    -------
+    float
+        weighted UniFrac distance between samples
+
+    Notes
+    -----
+    The count vectors passed here correspond to all nodes in the tree, not
+    just the tips as is the case for the public function
+    (``weighted_unifrac``).
+
+    """
+    if u_total_count:
+        u_ = u_node_counts / u_total_count
+    else:
+        u_ = 0.0
+
+    if v_total_count:
+        v_ = v_node_counts / v_total_count
+    else:
+        v_ = 0.0
+
+    return (branch_lengths * abs(u_ - v_)).sum()
+
+
+def _weighted_unifrac_normalized(u_node_counts, v_node_counts, u_total_count,
+                                 v_total_count, tree, tree_index):
+    """
+    Parameters
+    ----------
+    u_node_counts, v_node_counts : np.array
+         Vectors indicating presense (value greater than zero) and absense
+         (value equal to zero) of nodes in two samples, `u` and `v`. Order is
+         assumed to be the same as in `branch_lengths`.
+    u_total_count, v_total_counts : int
+         The sum of ``u_node_counts`` and ``v_node_counts`` vectors,
+         respectively. This could be computed internally, but since this is a
+         private method and the calling function has already generated these
+         values, this saves an iteration over each of these vectors.
+
+    Returns
+    -------
+    float
+        normalized weighted UniFrac distance between samples
 
+    Notes
+    -----
+    The count vectors passed here correspond to all nodes in the tree, not
+    just the tips as is the case for the public function
+    (``weighted_unifrac``).
+
+    """
+    branch_lengths = tree_index['length']
+    u = _weighted_unifrac(u_node_counts, v_node_counts, u_total_count,
+                          v_total_count, branch_lengths)
+    # get the index positions for tips in counts_array, and determine the
+    # tip distances to the root
+    tip_indices = np.array([n.id for n in tree_index['id_index'].values()
+                            if n.is_tip()])
+    node_to_root_distances = _tip_distances(branch_lengths, tree, tip_indices)
```

Comment:
@wasade, can this be computed as part of the index? Seems like computing this for every pair of samples will be a bottleneck.